segmatch/src/database.cpp

```diff
@@ -0,0 +1,389 @@
+    result << id_match_list_.at(i).at(j) << " ";
+    }
+    result << std::endl;
+  }
+  return result.str();
+}
+
+bool UniqueIdMatches::findId(const Id id, Position* position_ptr) const {
+  for (size_t i = 0u; i < id_match_list_.size(); ++i) {
+    for (size_t j = 0u; j < id_match_list_.at(i).size(); ++j) {
+      if (id == id_match_list_.at(i).at(j)) {
+        // If desired, pass the found id's position.
+        if (position_ptr != NULL) {
+          position_ptr->row = i;
+          position_ptr->col = j;
+        }
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+const std::string kDatabaseDirectory = "/tmp/segmatch/";
+const std::string kSegmentsFilename = "segments_database.csv";
+const std::string kFeaturesFilename = "features_database.csv";
+const std::string kMatchesFilename = "matches_database.csv";
+
+bool export_session_data_to_database(const SegmentedCloud& segmented_cloud,
+                                     const UniqueIdMatches& id_matches) {
+  return export_segments(kDatabaseDirectory + kSegmentsFilename, segmented_cloud) &&
+      export_features(kDatabaseDirectory + kFeaturesFilename, segmented_cloud) &&
+      export_matches(kDatabaseDirectory + kMatchesFilename, id_matches);
+}
+
+bool import_session_data_from_database(SegmentedCloud* segmented_cloud_ptr,
+                                       UniqueIdMatches* id_matches_ptr) {
+  return import_segments(kDatabaseDirectory + kSegmentsFilename, segmented_cloud_ptr) &&
+      import_features(kDatabaseDirectory + kFeaturesFilename, segmented_cloud_ptr) &&
+      import_matches(kDatabaseDirectory + kMatchesFilename, id_matches_ptr);
+}
+
+bool ensure_directory_exists(const std::string& directory) {
+  CHECK(directory.size() != 0u) << "Directory should not be an empty string.";
+  if (directory[0] == '/') {
+    boost::filesystem::path path(directory);
+    if(boost::filesystem::exists(path)) {
+      return true;
+    }
+    if(boost::filesystem::create_directory(path)) {
+      LOG(WARNING) << "Directory Created: " << directory;
+      return true;
+    }
+  } else {
+    LOG(ERROR) << "Directory '" << directory << "' starts with invalid character: '" <<
+        directory[0] << "'"; 
+  }
+  return false;
+}
+
+bool ensure_directory_exists_for_filename(const std::string& filename) {
+  size_t pos = filename.rfind("/");
+  if (pos != std::string::npos) {
+    std::string directory = filename.substr(0, pos);
+    return ensure_directory_exists(directory);
+  } else {
+    LOG(ERROR) << "Filename '" << filename << "' does not specify a directory.";
+    return false;
+  }
+}
+
+bool export_segments(const std::string& filename, const SegmentedCloud& segmented_cloud) {
+  ensure_directory_exists_for_filename(filename);
+  std::ofstream output_file;
+  output_file.open(filename, std::ofstream::out | std::ofstream::trunc);
+  if (output_file.is_open()) {
+    for (size_t i = 0u; i < segmented_cloud.getNumberOfValidSegments(); i++) {
```

Comment:
Here we could use `.begin()` and `.end()` which return iterators to the valid segments container